StockTrader/src/main/java/com/restResource/StockTrader/service/BuyTriggerService.java

```diff
@@ -0,0 +1,117 @@
+    private LoggingService loggingService;
+
+    public BuyTriggerService(LoggingService loggingService,
+                             QuoteService quoteService,
+                             BuyTriggerRepository buyTriggerRepository,
+                             TaskExecutor taskExecutor,
+                             InvestmentRepository investmentRepository,
+                             AccountRepository accountRepository){
+        this.quoteService = quoteService;
+        this.buyTriggerRepository = buyTriggerRepository;
+        this.taskExecutor = taskExecutor;
+        this.accountRepository = accountRepository;
+        this.investmentRepository = investmentRepository;
+        this.loggingService = loggingService;
+    }
+
+    public void start(String userId, String stockSymbol, Integer stockCost, Integer transactionNum) {
+        taskExecutor.execute(new Runnable() {
+            @Override
+            public void run(){
+                //wait for both rest calls to complete, verify sufficient funds.
+                for (;;) {
+
+                    try {
+                        Thread.sleep(5000);
+                    } catch(InterruptedException e) {
+                        throw new IllegalArgumentException(
+                                "Error with Thread.sleep in "+Thread.currentThread().getName());
+                    }
+
+                    Optional<BuyTrigger> buyStockSnapshot = buyTriggerRepository.findByUserIdAndStockSymbol(userId, stockSymbol);
+
+                    if (!buyStockSnapshot.isPresent()) { //we dont have an entry yet, so continue waiting
+                        continue;
+                    } else if (buyStockSnapshot.get().getStock_cost() != null) { //we already have a working buy trigger
+                        return;
+                    } else { //we have a trigger that is waiting for a cost target
+                        Integer amount = buyStockSnapshot.get().getStock_amount();
+                        if (accountRepository.removeFunds(userId, amount*stockCost, transactionNum, "TS1") == 0) {
+                            buyTriggerRepository.delete(buyStockSnapshot.get()); //remove it since it is not valid
+                            loggingService.logErrorEvent(
+                                    ErrorEventLog.builder()
+                                            .command(CommandType.SET_BUY_TRIGGER)
+                                            .username(userId)
+                                            .stockSymbol(stockSymbol)
+                                            .transactionNum(transactionNum)
+                                            .errorMessage("The amount parameter must be greater than zero")
+                                            .build());
+                            return;
+                        }
+                        buyStockSnapshot.get().setStock_cost(stockCost);
+                        buyTriggerRepository.save(buyStockSnapshot.get());
```

Comment:
This is a race condition. You need to have a proper lock here. One option would be to have separate tables for the trigger and amount and just attempt an insert on the trigger table. Feel free to come up with something else though.

Reply:
Could also just have a custom update query that checks cost is null

Reply:
This wouldnt be a race condition since its only possible to have one stockcost being set.

Reply:
If you sent out two SET_BUY_TRIGGER commands at the same time, they'd both check the cost and see that it was null, and then they would both save their cost, one of them overwriting the other. Or is there something else that is preventing multiple triggers from happening?

Reply:
Hmmm i see what you mean. I was thinking that the if would deal with it. I think i can wrap the whole thing in a synchronized statement

Reply:
You can also just have a query along the lines of UPDATE (cost) WHERE userId = some_user AND stock_symbol = some_stock AND cost = NULL; Then you just check if it updated.